### PR TITLE
[15.0][MIG] web_responsive: Migration to 15.0

### PR DESCRIPTION
**Pull request by SplashS**
Migration work has not been completed yet
But I share my work to get feedback and show cases to use OWL in v15

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Thank you very much!

**Comment by SplashS** [comment]
So I think I stabilized migration.
I have extra plan only to recapture screenshots. They are definitely outdated.

@pedrobaeza Some notes about hotkey overlays:

1. I left Alt modifier as in base odoo, but I added Alt + Shift modifier only for navbar menu entries.
This variant excludes conflict with Firefox tab switching. All current hotkeys works well in Chrome and Safari.
2. I tried to change Prev page and Next page hotkeys to Z and X accordingly. I think we should discuss it.

Also I made correct handling of window resize. Interface changed dynamically on device rotation f.x.

There are two implementations of ControlPanel. I was forced to patch both with code duplication. I wish in 16.0 there will be only one.

Waiting for feedback.

**Comment by SplashS** [comment]
@Tardo Migration is almost ready and runbot build named `runboat/build` works well.
This PR is ready for review.

File: web_responsive/static/src/components/navbar/main_navbar.esm.js
Comment (Tardo) [inline web_responsive/static/src/components/navbar/main_navbar.esm.js]: Why is this necessary?

**Review by Tardo — COMMENTED** [review]
Thanks, good job! I have seen some little "glitches":
- Tree Kanban doesn't look as in previous versions:
responsive_15_kanban_tree

- The attachment zone of the chatter should be fixed when use the scroll:
responsive_15_attachments_zone

It may also come in handy to refactor to change the use of "var" to let or const, avoid the use of "self" by using "arrow functions", etc..

File: web_responsive/static/src/components/navbar/main_navbar.esm.js
Comment (SplashS) [reply web_responsive/static/src/components/navbar/main_navbar.esm.js]: It was necessary before https://github.com/odoo/odoo/commit/3b18f1cf8cfdbc497c8acba67970e835dabc2eb2#diff-c8d085006528c9e527b7d7575894d1ab86722180c60be7a75f50187ab6c5d14f
Now it is unnecessary. I will remove it

Thanks for feedback! I will consider it soon and let you now

**Comment by SplashS** [comment]
@Tardo 
> * Tree Kanban doesn't look as in previous versions:

Fixed

> * The attachment zone of the chatter should be fixed when use the scroll:

I looked at 14.0 version. Attachments zone isn't fixed. In my opinion it's dangerous to make it fixed because attachments count isn't limited. 

> It may also come in handy to refactor to change the use of "var" to let or const, avoid the use of "self" by using "arrow functions", etc..

Done

**Comment by jcdrubay** [comment]
Tested in local and I could not see the sub menus

I could not test on runbot nor runboat, maybe a new build needs to be triggered by a new commit ?

image

**Comment by SplashS** [comment]
> Tested in local and I could not see the sub menus
> 
> I could not test on runbot nor runboat, maybe a new build needs to be triggered by a new commit ?


Thank you very much for feedback.
It's broken since https://github.com/odoo/odoo/commit/3659d65f37d08737f99d62ecb8cd486ba8d764f4
Fixed.

Runboat will work few days after my push.

**Comment by jcdrubay** [comment]
Thanks @SplashS 
Runboat is now working well and I could see the sub-menus nicely 👍

**Review by jcdrubay — APPROVED** [review]
LGTM after running few tests on Runboat. Thanks @SplashS

**Comment by stfnndrsn** [comment]
I just tried to install `web_responsive` on a clean v15  database; but I get the following exception

```RPC_ERROR
Odoo Server Error
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/tools/cache.py", line 85, in lookup
    r = d[key]
  File "/usr/lib/python3/dist-packages/odoo/tools/func.py", line 71, in wrapper
    return func(self, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/tools/lru.py", line 34, in __getitem__
    a = self.d[obj]
KeyError: ('ir.model.data', <function IrModelData._xmlid_lookup at 0x7f2c710e8dc0>, 'web.assets_backend')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/tools/convert.py", line 680, in _tag_root
    f(rec)
  File "/usr/lib/python3/dist-packages/odoo/tools/convert.py", line 658, in _tag_template
    return self._tag_record(record)
  File "/usr/lib/python3/dist-packages/odoo/tools/convert.py", line 567, in _tag_record
    f_val = self.id_get(f_ref)
  File "/usr/lib/python3/dist-packages/odoo/tools/convert.py", line 663, in id_get
    res = self.model_id_get(id_str, raise_if_not_found)
  File "/usr/lib/python3/dist-packages/odoo/tools/convert.py", line 669, in model_id_get
    return self.env['ir.model.data']._xmlid_to_res_model_res_id(id_str, raise_if_not_found=raise_if_not_found)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/models/ir_model.py", line 1935, in _xmlid_to_res_model_res_id
    return self._xmlid_lookup(xmlid)[1:3]
  File "<decorator-gen-35>", line 2, in _xmlid_lookup
  File "/usr/lib/python3/dist-packages/odoo/tools/cache.py", line 90, in lookup
    value = d[key] = self.method(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/models/ir_model.py", line 1928, in _xmlid_lookup
    raise ValueError('External ID not found in the system: %s' % xmlid)
ValueError: External ID not found in the system: web.assets_backend

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/addons/base/models/ir_http.py", line 237, in _dispatch
    result = request.dispatch()
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 688, in dispatch
    result = self._call_function(**self.params)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 360, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 349, in checked_call
    result = self.endpoint(*a, **kw)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 917, in __call__
    return self.method(*args, **kw)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 536, in response_wrap
    response = f(*args, **kw)
  File "/usr/lib/python3/dist-packages/odoo/addons/web/controllers/main.py", line 1339, in call_button
    action = self._call_kw(model, method, args, kwargs)
  File "/usr/lib/python3/dist-packages/odoo/addons/web/controllers/main.py", line 1327, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 464, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 451, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "<decorator-gen-74>", line 2, in button_immediate_install
  File "/usr/lib/python3/dist-packages/odoo/addons/base/models/ir_module.py", line 74, in check_and_log
    return method(self, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/models/ir_module.py", line 483, in button_immediate_install
    return self._button_immediate_function(type(self).button_install)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/models/ir_module.py", line 600, in _button_immediate_function
    registry = modules.registry.Registry.new(self._cr.dbname, update_module=True)
  File "/usr/lib/python3/dist-packages/odoo/modules/registry.py", line 87, in new
    odoo.modules.load_modules(registry, force_demo, status, update_module)
  File "/usr/lib/python3/dist-packages/odoo/modules/loading.py", line 474, in load_modules
    processed_modules += load_marked_modules(cr, graph,
  File "/usr/lib/python3/dist-packages/odoo/modules/loading.py", line 363, in load_marked_modules
    loaded, processed = load_module_graph(
  File "/usr/lib/python3/dist-packages/odoo/modules/loading.py", line 222, in load_module_graph
    load_data(cr, idref, mode, kind='data', package=package)
  File "/usr/lib/python3/dist-packages/odoo/modules/loading.py", line 69, in load_data
    tools.convert_file(cr, package.name, filename, idref, mode, noupdate, kind)
  File "/usr/lib/python3/dist-packages/odoo/tools/convert.py", line 745, in convert_file
    convert_xml_import(cr, module, fp, idref, mode, noupdate)
  File "/usr/lib/python3/dist-packages/odoo/tools/convert.py", line 811, in convert_xml_import
    obj.parse(doc.getroot())
  File "/usr/lib/python3/dist-packages/odoo/tools/convert.py", line 731, in parse
    self._tag_root(de)
  File "/usr/lib/python3/dist-packages/odoo/tools/convert.py", line 693, in _tag_root
    raise ParseError('while parsing %s:%s, somewhere inside\n%s' % (
Exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 644, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 302, in _handle_exception
    raise exception.with_traceback(None) from new_cause
odoo.tools.convert.ParseError: while parsing None:12, somewhere inside
<data name="Open Mobile Assets" inherit_id="web.assets_backend">
        <xpath expr=".">
            <link rel="stylesheet" href="/web_responsive/static/src/css/web_responsive.scss"/>
            <script type="application/javascript" src="/web_responsive/static/src/js/web_responsive.js"/>
            <script type="application/javascript" src="/web_responsive/static/src/js/discuss.js"/>
        </xpath>
    </data>
```

**Review by Tardo — CHANGES_REQUESTED** [review]
Overall it is a great job ;)

I get this exception when resizing. I load the page with small size and stretch it:
web_responsive_exception_1

Please, update the readme with the explanation of how shortcuts works...

Please apply/adapt this commits: 622991abe3fa56013d0ec36292092f3ebf19d554 and 06269897f130764f2802225f8eda5e02a9cb160b

**Comment by SplashS** [comment]
> I get this exception when resizing. I load the page with small size and stretch it

Yes, it's the big problem for me. Exception caused in my code that should rerender kanban one2many as list on resize.
But it's initialization ties with action initialization and I still don't have solution to rerender one2many on the fly.
Will try one more time. Anyway any suggestions welcome.

**Comment by SplashS** [comment]
> I get this exception when resizing. I load the page with small size and stretch it:

I fixed it but kanban/list mode doesn't switch automatically for X2Many fields. Anyway it works.

> Please, update the readme with the explanation of how shortcuts works...

If I understand right mergebot will update README from readme/DESCRIPTION. Here I updated shortcuts explanation.

> Please apply/adapt this commits: [622991a](https://github.com/OCA/web/commit/622991abe3fa56013d0ec36292092f3ebf19d554) and [0626989](https://github.com/OCA/web/commit/06269897f130764f2802225f8eda5e02a9cb160b)

Done

**Review by Tardo — APPROVED** [review]
Thanks for the great work!

**Comment by pedrobaeza** [comment]
@SplashS one last functional tests by my part, with a couple of things:

- On the AppSwitcher, if you focus on the search textbox, pressing Esc doesn't close it as before.
- In combination with `web_environment_ribbon`,  the ribbon disappears when hovering on the top bar:

  Peek 25-01-2022 14-13

  Is it a problem of this module or the other?

Finally, I think your company deserves a place in the author list, due to all your work in the latest versions.

**Comment by SplashS** [comment]
> * On the AppSwitcher, if you focus on the search textbox, pressing Esc doesn't close it as before.

Yep! Not the simplest one solution, but it's fixed.

> * In combination with `web_environment_ribbon`,  the ribbon disappears when hovering on the top bar:
>   Is it a problem of this module or the other?

I removed `web_responsive` from apps and `web_environment_ribbon` behavior is the same. So I think it's on ribbon module side. Quick look gives me no idea why. But it disappears on hovering navbar in any point.

> Finally, I think your company deserves a place in the author list, due to all your work in the latest versions.

Thank you, I've forgotten about this place.

**Comment by jcdrubay** [comment]
Thank you @SplashS and all who supported for this great migration :)

**Pull request merged**
